src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -145,6 +154,148 @@ public NetworkPeerClient(int Id, TcpClient tcpClient, ILoggerFactory loggerFacto
             this.logger.LogTrace("(-)");
         }
 
+        /// <summary>
+        /// Reads raw message in binary form from the connection stream.
+        /// </summary>
+        /// <param name="protocolVersion">Version of the protocol that defines the message format.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the read operation.</param>
+        /// <returns>Binary message received from the connected counterparty.</returns>
+        /// <exception cref="OperationCanceledException">Thrown if the operation was cancelled or the end of the stream was reached.</exception>
+        /// <exception cref="FormatException">Thrown if the incoming message is too big.</exception>
+        public async Task<byte[]> ReadMessageAsync(ProtocolVersion protocolVersion, CancellationToken cancellation = default(CancellationToken))
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(protocolVersion), protocolVersion);
+
+            // First find and read the magic.
+            await this.ReadMagicAsync(this.network.MagicBytes, cancellation).ConfigureAwait(false);
+
+            // Then read the header, which is formed of command, length, and possibly also a checksum.
+            int checksumSize = protocolVersion >= ProtocolVersion.MEMPOOL_GD_VERSION ? Message.ChecksumSize : 0;
+            int headerSize = Message.CommandSize + Message.LengthSize + checksumSize;
+
+            byte[] messageHeader = new byte[headerSize];
+            await this.ReadBytesAsync(messageHeader, 0, headerSize, cancellation).ConfigureAwait(false);
+
+            // Then extract the length, which is the message payload size.
+            int lengthOffset = Message.CommandSize;
+            uint length = BitConverter.ToUInt32(messageHeader, lengthOffset);
+
+            // 32 MB limit on message size from Bitcoin Core.
+            if (length > 0x02000000)
+                throw new FormatException("Message payload too big (over 0x02000000 bytes)");
```

Comment:
`FormatException`? maybe we need a `ProtocolViolationException`

Reply:
ok i can change it in later PR